osu.Game.Rulesets.Osu/Mods/OsuModRelax.cs

```diff
@@ -33,13 +34,15 @@ public class OsuModRelax : ModRelax, IUpdatableByPlayfield, IApplicableToDrawabl
 
         private ReplayState<OsuAction> state;
         private double lastStateChangeTime;
+        private double beatmapLength;
 
         private bool hasReplay;
 
         public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
         {
             // grab the input manager for future use.
             osuInputManager = (OsuInputManager)drawableRuleset.KeyBindingInputManager;
+            beatmapLength = drawableRuleset.Beatmap.HitObjects.LastOrDefault()?.GetEndTime() ?? 0;
```

Comment:
Will this ever be null?

Reply:
You can easily check that yourself by exporting a zero-objects map and apply the mod and try it, but either way it shouldn't and will at least have one.

Reply:
This check is (theoretically) wrong. The objects being ordered by their *start* time does *not* mean that the *last* object sorted by start time has the greatest *end* time. See cases with overlapping objects.

Reply:
Yeah this is already the case on `SongProgress`:
https://github.com/ppy/osu/blob/dc50ae40b9b604785c1093c3c5cdc9dbb6160876/osu.Game/Screens/Play/SongProgress.cs#L54-L55

Which opted out to a TODO comment instead.

Also, I think using the name `lastHitTime` here is much better than `beatmapLength`.

Reply:
So is this ok?
`drawableRuleset.Beatmap.HitObjects.OrderBy(h => h.GetEndTime()).LastOrDefault().GetEndTime()`

Reply:
`drawableRuleset.Objects.Max(h => h.GetEndTime);`